### DISCUSSION TITLE
How to chain atomsWithQuery calls together

### DISCUSSION THREAD

**Discussion by sazzer** [pmndrs/jotai#1998] [Q&A]
I'm writing an app built around a hypermedia API. As such, in order to access a resource it involves following links from other resources.

For example, to start with I'm loading the details of the current user. The process for this is:
* Load the API home document - `GET /`.
* If there is an "authenticated-as" link in this, follow that to get the current user.
* If there isn't an "authenticated-as" link then we're not authenticated and so there is no current user.

Seems simple enough.

So, I've written an `atomsWithQuery` for the API home document:
```
export const [homeAtom] = atomsWithQuery<HAL<HomeDocument>>(() => {
  return {
    queryKey: ['home'],
    queryFn: async () => {
      const res = await fetch('/');
      const body = await res.json();

      LOG('Home document', { body });

      return body;
    }
  };
});
```

I've then written an atom to extract the "authenticated-as" link from the home document:
```
const currentUserIdAtom = atom<Promise<string | undefined>>(async (get) => {
  const home = await get(homeAtom);

  const authenticatedAsLink = findLink(home, 'authenticated-as');

  LOG('Current user id', { authenticatedAsLink });

  return authenticatedAsLink?.href;
});
```

So far, that all works. Though it is doing excessive work for some reason, but that's probably another problem for another day :)
<img width="607" alt="image" src="https://github.com/pmndrs/jotai/assets/66072/57e0b3ce-3ca1-46ea-a30d-1576691dbf2e">

Next I've written an `atomsWithQuery` for the current user, based off of this one:
```
export const [currentUserAtom] = atomsWithQuery((get) => ({
  queryKey: ['home', get(currentUserIdAtom)],
  queryFn: async ({ queryKey: [, id] }) => {
    LOG('Getting user', { id });

    if (id === undefined) {
      return;
    }

    const res = await fetch(id);
    const body = await res.json();

    LOG('User document', { body });

    return body;
  }
}));
```

And this *doesn't* work. The `id` value in my `queryFn` is always a `Promise` and not the actual value, which means that my fetch calls don't work correctly:
<img width="829" alt="image" src="https://github.com/pmndrs/jotai/assets/66072/fe2f85f3-0991-4aec-9765-cff51543d01a">
<img width="523" alt="image" src="https://github.com/pmndrs/jotai/assets/66072/a584309e-b7a4-4f4c-9922-fe8cafe6df14">

Is there something I'm missing here to be able to call one `atomsWithQuery` with the result of another one?

Cheers

**Comment by dai-shi** [comment]
In general, I don't think it's a good idea to chain atomWithQuery. It's far more efficient to have one combined async function and one atom, if possible.

For async params, we provide another function: https://github.com/jotaijs/jotai-tanstack-query/issues/21 and https://github.com/jotaijs/jotai-tanstack-query/pull/30
